Sources/MongoSwift/BSON/BSON.swift

```diff
@@ -0,0 +1,413 @@
+    public func asInt() -> Int? {
+        switch self {
+        case let .int32(value):
+            return Int(value)
+        case let .int64(value):
+            return Int(exactly: value)
+        case let .double(value):
+            return Int(exactly: value)
+        default:
+            return nil
+        }
+    }
+
+    /// Return this BSON as an `Int32` if possible.
+    /// This will coerce numeric cases (e.g. `.double`) into an `Int32` if such coercion would be lossless.
+    public func asInt32() -> Int32? {
+        switch self {
+        case let .int32(value):
+            return value
+        case let .int64(value):
+            return Int32(exactly: value)
+        case let .double(value):
+            return Int32(exactly: value)
+        default:
+            return nil
+        }
+    }
+
+    /// Return this BSON as an `Int64` if possible.
+    /// This will coerce numeric cases (e.g. `.double`) into an `Int64` if such coercion would be lossless.
+    public func asInt64() -> Int64? {
+        switch self {
+        case let .int32(value):
+            return Int64(value)
+        case let .int64(value):
+            return value
+        case let .double(value):
+            return Int64(exactly: value)
+        default:
+            return nil
+        }
+    }
+
+    /// Return this BSON as a `Double` if possible.
+    /// This will coerce numeric cases (e.g. `.decimal128`) into a `Double` if such coercion would be lossless.
+    public func asDouble() -> Double? {
+        switch self {
+        case let .double(d):
+            return d
+        default:
+            guard let intValue = self.asInt() else {
+                return nil
+            }
+            return Double(intValue)
+        }
+    }
+
+    /// Return this BSON as a `Decimal128` if possible.
+    /// This will coerce numeric cases (e.g. `.double`) into a `Decimal128` if such coercion would be lossless.
+    public func asDecimal128() -> Decimal128? {
+        switch self {
+        case let .decimal128(d):
+            return d
+        case let .int64(i):
+            return Decimal128(String(i))
+        case let .int32(i):
+            return Decimal128(String(i))
+        case let .double(d):
+            return Decimal128(String(d))
+        default:
+            return nil
+        }
+    }
+
+    /// Return this BSON as an `[T]` if possible, transforming each element via the provided closure.
+    /// If the closure returns nil, the value will be discarded from the resultant array.
+    public func asArray<T>(_ transform: (BSON) -> T?) -> [T]? {
+        return self.arrayValue?.compactMap(transform)
+    }
+}
+
+/// Extension providing the internal API of `BSON`
+extension BSON {
+    /// List of all BSONValue types. Can be used to exhaustively check each one at runtime.
+    internal static var allBSONTypes: [BSONValue.Type] = [
```

Comment:
this will be used by the `Decodable` conformance to try a bunch of different types.

Reply:
i never thought about this in the design, but what happens to the types like null, undefined, minkey, maxkey now, since they aren't actually used in the enum?

Reply:
I was going to expand on this more in the next PR, but I think we can at the least internalize those types (and then remove them after that if need be). If I remember right, we were still using them in the pure swift bson implementation to encode to bytes and what not.